t/Mailbox/query/basic.t

```diff
@@ -0,0 +1,485 @@
+      my $res = $tester->request({
+        using => [ "ietf:jmapmail" ],
+        methodCalls => [[
+          "Mailbox/query" => {
+            filter => {
+              parentId => undef,
+            },
+          },
+        ]],
+      });
+      ok($res->is_success, "Mailbox/query")
+        or diag explain $res->http_response->as_string;
+
+      my $ids = $res->single_sentence("Mailbox/query")->arguments->{ids};
+
+      jcmp_deeply($ids, supersetof($mailbox1->id), 'Got top-level mailbox');
+      jcmp_deeply($ids, noneof($mailbox2->id), 'Did not get sub-mailbox');
+    };
+
+    subtest "has a parentId" => sub {
+      my $res = $tester->request({
+        using => [ "ietf:jmapmail" ],
+        methodCalls => [[
+          "Mailbox/query" => {
+            filter => {
+              parentId => $mailbox1->id,
+            },
+          },
+        ]],
+      });
+      ok($res->is_success, "Mailbox/query")
+        or diag explain $res->http_response->as_string;
+
+      my $ids = $res->single_sentence("Mailbox/query")->arguments->{ids};
+
+      jcmp_deeply($ids, [ $mailbox2->id ], 'Got our one child mailbox');
+    };
+  };
+
+  subtest "hasRole" => sub {
+    # Find some mailboxes with roles
+    my $res = $tester->request({
+      using => [ "ietf:jmapmail" ],
+      methodCalls => [[
+        "Mailbox/get" => {},
+      ]],
+    });
+
+    my @with_roles = map {;
+      $_->{id}
+    } grep {;
+      $_->{role}
+    } @{ $res->single_sentence("Mailbox/get")->arguments->{list} };
+
+    plan skip_all => "No mailboxes with roles found, can't continue"
+      unless @with_roles;
+
+    subtest "false" => sub {
+      my $res = $tester->request({
+        using => [ "ietf:jmapmail" ],
+        methodCalls => [[
+          "Mailbox/query" => {
+            filter => {
+              hasRole => JSON::false,
+            },
+          },
+        ]],
+      });
+      ok($res->is_success, "Mailbox/query")
+        or diag explain $res->http_response->as_string;
+
+      my $ids = $res->single_sentence("Mailbox/query")->arguments->{ids};
+
+      jcmp_deeply(
+        $ids,
+        supersetof($mailbox1->id, $mailbox2->id),
+        'Got mailboxes with no roles',
+      );
+
+      jcmp_deeply(
+        $ids,
+        noneof(@with_roles),
+        'Did not get mailboxes with roles',
+      ) or diag explain $ids;
+    };
+
+    subtest "true" => sub {
+      my $res = $tester->request({
+        using => [ "ietf:jmapmail" ],
+        methodCalls => [[
+          "Mailbox/query" => {
+            filter => {
+              hasRole => JSON::true,
+            },
+          },
+        ]],
+      });
+      ok($res->is_success, "Mailbox/query")
+        or diag explain $res->http_response->as_string;
+
+      my $ids = $res->single_sentence("Mailbox/query")->arguments->{ids};
+
+      jcmp_deeply($ids, \@with_roles, 'Got mailboxes with roles only');
+    };
+  };
+};
+
+pristine_test "sorting and limiting" => sub {
+  my ($self) = @_;
```

Comment:
This whole thing feels ripe for abstraction for future tests.

Reply:
Absolutely. Waiting on a second set so I can better figure out what that looks lik